src/core/device.c

```diff
@@ -158,19 +158,65 @@ static void device_set_state(Device *d, DeviceState state) {
+        if (d->deserialized_state < 0)
+                return 0;
+
+        m = u->manager;
+        f = d->deserialized_found;
+        s = d->deserialized_state;
+
+        /* On initial boot, switch-root, reload, reexecute, the following happen:
+         * 1. MANAGER_IS_RUNNING() == false
+         * 2. enumerate devices: manager_enumerate() -> device_enumerate()
+         *    Device::enumerated_found is set.
+         * 3. deserialize devices: manager_deserialize() -> device_deserialize()
+         *    Device::deserialize_state and Device::deserialized_found are set.
+         * 4. coldplug devices: manager_coldplug() -> device_coldplug()
+         *    deserialized properties are copied to the main properties.
+         * 5. MANAGER_IS_RUNNING() == true: manager_ready()
+         * 6. catchup devices: manager_catchup() -> device_catchup()
+         *    Device::enumerated_found is applied to Device::found, and state is updated based on that.
+         *
+         * Notes:
+         * - On initial boot, no udev database exist. Hence, no devices are enumerated in the step 2.
+         *   Also, there is no deserialized device. Device units are (a) generated based on dependencies of
+         *   other units, or (b) generated when uevents are received.
+         *
+         * - On switch-root, no udev databse exist, except for devices with sticky bit, i.e. OPTIONS="db_persist".
+         *   Hence, almost no devices are enumerated in the step 2. However, in general, we have several
+         *   serialized devices. But, DEVICE_FOUND_UDEV bit in the deserialized_found must be ignored, as udev
+         *   rules in initramfs and the main system are often different. Hence, if the deserialized state is
+         *   DEVICE_PLUGGED, we need to downgrade it to DEVICE_TENTATIVE (or DEVICE_DEAD if nobody sees the device).
+         *   Unlike the other starting mode, Manager::honor_device_enumeration == false (maybe, it is better to
+         *   rename the flag) when device_coldplug() and device_catchup() are called. Hence, let's conditionalize
+         *   the operations by using the flag. After switch-root, systemd-udevd will (re-)process all devices,
+         *   and the Device::found and Device::state will be adjusted.
+         *
+         * - On reload or reexecute, we can trust enumerated_found, deserialized_found, and deserialized_state.
+         *   Of course, deserialized parameters may be outdated, but the unit state can be adjusted later by
+         *   device_catchup() or uevents. */
+
+        if (!m->honor_device_enumeration && !MANAGER_IS_USER(m)) {
+                f &= ~DEVICE_FOUND_UDEV; /* ignore DEVICE_FOUND_UDEV bit */
+                if (s == DEVICE_PLUGGED)
+                        s = DEVICE_TENTATIVE; /* downgrade state */
+                if (f == DEVICE_NOT_FOUND)
+                        s = DEVICE_DEAD; /* If nobody sees the device, downgrade more */
```

Comment:
I assume that `deserialized_found == DEVICE_NOT_FOUND` implies `deserialized_state == DEVICE_DEAD`. In that case the clause would be redundant. But if this implication does **not** hold for whatever reason, we'd better understand how that comes to pass before we reset the state to "dead", which is dangerous.

Reply:
It would be strange to find this combination (`found == DEVICE_NOT_FOUND` and `state != DEVICE_DEAD`) in the serialized state, yes. Maybe we should emit a warning if we hit this case.

Reply:
OK. Let's add the warning later.